### DISCUSSION TITLE
Meter Tags and Prometheus HttpListener

### DISCUSSION THREAD

**Discussion by ProbablePrime** [open-telemetry/opentelemetry-dotnet#5829] [Q&A] [answered]
Hello, 

I'm having difficult understanding, the use of Tags and the resultant Prometheus output, from a `System.Diagnostics.Metrics.Meter` vs a `System.Diagnostics.Metrics.Instrument<T>`'s Add/Set Method.

The [documentation for Metric Enrichment](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics#metrics-enrichment) mentions:
> Additional tags provided at meter creation time. For example, the [Meter(name, version, tags, scope)](https://learn.microsoft.com/dotnet/api/system.diagnostics.metrics.meter.-ctor) overload.

But these do not appear to be working, when tags from Meter creation are specified they are not present in the Prometheus Output.

I started looking over on Stack Overflow, and found [this question](https://stackoverflow.com/questions/78806211/how-to-have-the-prometheus-exporter-export-labels-tags). Which is identical to my issue. 

But I wanted to make it even easier to help the community here see what's going on.

So on a fork of opentelemetry-dotnet, I have edited a test to show the issue. You can find the [whole commit](https://github.com/open-telemetry/opentelemetry-dotnet/commit/e4c7d2b7b7fbb4d8e64a08a8d2ae8473bfa9a8ec).

But, to help further:
1. Here's where I have added [Meter scoped tags](https://github.com/open-telemetry/opentelemetry-dotnet/commit/e4c7d2b7b7fbb4d8e64a08a8d2ae8473bfa9a8ec#diff-ad457b7cc11412d4d6339428461e04c3938d7614391fe2f3dc72d51946002cbfR243-R249)
2. [Here's where I edited the test output, to include the new meter tags](https://github.com/open-telemetry/opentelemetry-dotnet/commit/e4c7d2b7b7fbb4d8e64a08a8d2ae8473bfa9a8ec#diff-ad457b7cc11412d4d6339428461e04c3938d7614391fe2f3dc72d51946002cbfR300)

These changes result in a failing test as the Meter based tags are not present, but based on the documentation, I feel this should be working.

It looks like only tags that are sent with a metric's value (via Set, Add, Record etc) are propagated.

However, questions on this area seem to come up somewhat frequently so I'm a little confused. For example there's:
- https://github.com/open-telemetry/opentelemetry-dotnet/issues/5478
- https://github.com/open-telemetry/opentelemetry-dotnet/discussions/5313
- https://github.com/open-telemetry/opentelemetry-dotnet/pull/5089  - This one introduced the test I modified.

As such I am not sure:
- If this is a new issue/bug that needs to be reported as an Issue?
- If this is covered elsewhere such as the items I linked
- If this is not possible due to another factor.

Could anyone provide some guidance?

**Comment by ProbablePrime** [comment]
This also affects "Observable" Instruments.

`Capacity = Meter.CreateObservableGauge<int>($"{MetricPrefix}.capacity", () => capacity, "XXX", "XXX", Tags);`

Tags created in this way also do not show up.

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus.AspNetCore/CHANGELOG.md#180-beta1

Prometheus supports instrumentation scope (Meter tags), only when OpenMetrics format is used. Is that your case?

**Reply by ProbablePrime** [reply]
Thank you!, I was not using the "application/openmetrics-text", content header in my target application. 

However, even when swapping to it, Meter Tags still do not show up.

I updated the test case fixture on my fork, to only have the meter tags in the requestOpenMetrics version of the fixture: https://github.com/ProbablePrime/opentelemetry-dotnet/commit/69394b5977ab2c8dfa32947cdff779f91a6fd0be

This test still fails: 
![image](https://github.com/user-attachments/assets/5a89587f-0874-44b9-9d8c-4c1747d2fbb9)

**Reply by cijothomas** [reply, marked as answer]
Got it. It looks like a bug in Prometheus exporter. It is only exposing Meter name for the scope, not the version, not the tags! Could you report it as a bug?

**Reply by ProbablePrime** [reply]
I will get to this as soon as I can. I'll see about making a minimum reproduction to help as well!

**Reply by cijothomas** [reply]
It is well understood, so no need to work on a repro.

@robertcoltheart Would you have some bandwidth to make this feature?

**Reply by robertcoltheart** [reply]
Sure, will take a look this week 👍

**Reply by ProbablePrime** [reply]
Ok I'll stand up a bug today!

Thank you for this!

**Reply by ProbablePrime** [reply]
Opened at: https://github.com/open-telemetry/opentelemetry-dotnet/issues/5836

**Reply by robertcoltheart** [reply]
Added https://github.com/open-telemetry/opentelemetry-dotnet/pull/5837 for review
